src/value.zig

```diff
@@ -1233,15 +1233,161 @@ pub const Value = extern union {
     }
 
     pub fn eql(a: Value, b: Value) bool {
-        if (a.tag() == b.tag() and a.tag() == .enum_literal) {
-            const a_name = @fieldParentPtr(Payload.Bytes, "base", a.ptr_otherwise).data;
-            const b_name = @fieldParentPtr(Payload.Bytes, "base", b.ptr_otherwise).data;
-            return std.mem.eql(u8, a_name, b_name);
+        if (a.tag() == b.tag()) {
+            if (a.tag() == .void_value or a.tag() == .null_value) {
+                return true;
+            } else if (a.tag() == .enum_literal) {
+                const a_name = @fieldParentPtr(Payload.Bytes, "base", a.ptr_otherwise).data;
+                const b_name = @fieldParentPtr(Payload.Bytes, "base", b.ptr_otherwise).data;
+                return std.mem.eql(u8, a_name, b_name);
+            }
+        }
+        if (a.isType() and b.isType()) {
+            // 128 bytes should be enough to hold both types
+            var buf: [128]u8 = undefined;
+            var fib = std.heap.FixedBufferAllocator.init(&buf);
+            const a_type = a.toType(&fib.allocator) catch unreachable;
+            const b_type = b.toType(&fib.allocator) catch unreachable;
+            return a_type.eql(b_type);
         }
-        // TODO non numerical comparisons
         return compare(a, .eq, b);
     }
 
+    pub fn hash(self: Value) u64 {
```

Comment:
Note that this function must return the same hash value for the same values even when they are represented differently. For example, with this implementation, .zero, .int_u64, .int_i64, .big_int_positive, and .big_int_negative are all capable of representing the value `0` but they will incorrectly give different hash values.